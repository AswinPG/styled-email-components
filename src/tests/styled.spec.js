import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import styled, { css } from '../';

const render = element => ReactTestRenderer.create(element).toJSON();

describe('styled', () => {
  it('works with `styled.component`', () => {
    const Link = styled.a`
      font-weight: bold;
      color: blue;
    `;
    expect(render(<Link>Hey</Link>)).toMatchSnapshot();
  });

  it('works with `styled(Component)`', () => {
    const Link = styled(props => <a {...props} />)`
      font-weight: bold;
      color: green;
    `;
    expect(render(<Link>Hey</Link>)).toMatchSnapshot();
  });

  it('interpolates props-based values correctly', () => {
    const Text = styled.p`
      color: #333;
      font-size: ${props => (props.big ? 32 : 16)}px;
    `;
    expect(render(<Text big>Hey</Text>)).toMatchSnapshot();
  });

  it('composes components', () => {
    const Link = styled.a`
      text-decoration: underline;
      font-size: 16px;
      color: blue;
    `;
    const Button = styled(Link)`
      text-decoration: none;
      font-weight: bold;
      padding: 5px 10px;
      background: blue;
      color: white;
    `;
    const BigButton = styled(Button)`
      padding: 10px 20px;
      font-size: 20px;
    `;
    expect(
      render(<BigButton href="https://example.com">Click me</BigButton>),
    ).toMatchSnapshot();
  });

  it('works with `.attrs`', () => {
    const Action = styled.a.attrs({
      href: 'https://example.com',
      target: 'blank',
    })`
      color: blue;
    `;
    expect(render(<Action>Click me</Action>)).toMatchSnapshot();
  });

  it('works with `.extend`', () => {
    const Text = styled.p`
      font-size: 16px;
      color: #333;
    `;
    const BigText = Text.extend`
      font-size: 32px;
    `;
    expect(render(<BigText>Hey</BigText>)).toMatchSnapshot();
  });

  it('works with `.withComponent`', () => {
    const Text = styled.p`
      color: green;
    `;
    const Heading = Text.withComponent('h1');
    expect(render(<Heading>Hey</Heading>)).toMatchSnapshot();
  });

  it('works with `css` helper', () => {
    const mixin = css`
      color: ${props => (props.primary ? 'blue' : '#333')};
    `;
    const Link = styled.a`
      font-size: 16px;
      ${mixin};
    `;
    expect(render(<Link primary>Click</Link>)).toMatchSnapshot();
  });

  it('warns about nested rules', () => {
    let lastConsoleWarn;
    const originalConsoleWarn = console.warn;
    console.warn = (...message) => (lastConsoleWarn = message.join(' '));
    const Invalid = styled.div`
      color: #333;

      // This would't work
      .nested {
        color: blue;
      }
    `;
    expect(render(<Invalid />)).toMatchSnapshot();
    expect(lastConsoleWarn).toMatchSnapshot();
    console.warn = originalConsoleWarn;
  });

  it('expands style rules', () => {
    const Box = styled.p`
      border: 1px solid #333;
      background: #333;
    `;
    expect(render(<Box />)).toMatchSnapshot();
  });

  it('merges with `style` prop', () => {
    const Link = styled.a`
      font-size: 16px;
      color: red;
    `;
    expect(render(<Link style={{ color: 'blue' }} />)).toMatchSnapshot();
  });
});
