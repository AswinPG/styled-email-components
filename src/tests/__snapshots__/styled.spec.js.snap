// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled composes components 1`] = `
<a
  href="https://example.com"
  style={
    Object {
      "backgroundColor": "blue",
      "color": "white",
      "fontSize": "20px",
      "fontWeight": "bold",
      "paddingBottom": "10px",
      "paddingLeft": "20px",
      "paddingRight": "20px",
      "paddingTop": "10px",
      "textDecorationColor": "black",
      "textDecorationLine": "none",
      "textDecorationStyle": "solid",
    }
  }
>
  Click me
</a>
`;

exports[`styled expands style rules 1`] = `
<p
  style={
    Object {
      "backgroundColor": "#333333",
      "borderColor": "#333333",
      "borderStyle": "solid",
      "borderWidth": "1px",
    }
  }
/>
`;

exports[`styled interpolates props-based values correctly 1`] = `
<p
  style={
    Object {
      "color": "#333333",
      "fontSize": "32px",
    }
  }
>
  Hey
</p>
`;

exports[`styled merges with \`style\` prop 1`] = `
<a
  style={
    Object {
      "color": "blue",
      "fontSize": "16px",
    }
  }
/>
`;

exports[`styled warns about nested rules 1`] = `
<div
  style={
    Object {
      "color": "#333333",
    }
  }
/>
`;

exports[`styled warns about nested rules 2`] = `"Node of type rule not supported as an inline style"`;

exports[`styled works with \`.attrs\` 1`] = `
<a
  href="https://example.com"
  style={
    Object {
      "color": "blue",
    }
  }
  target="blank"
>
  Click me
</a>
`;

exports[`styled works with \`.extend\` 1`] = `
<p
  style={
    Object {
      "color": "#333333",
      "fontSize": "32px",
    }
  }
>
  Hey
</p>
`;

exports[`styled works with \`.withComponent\` 1`] = `
<h1
  style={
    Object {
      "color": "green",
    }
  }
>
  Hey
</h1>
`;

exports[`styled works with \`css\` helper 1`] = `
<a
  style={
    Object {
      "color": "blue",
      "fontSize": "16px",
    }
  }
>
  Click
</a>
`;

exports[`styled works with \`styled(Component)\` 1`] = `
<a
  style={
    Object {
      "color": "green",
      "fontWeight": "bold",
    }
  }
>
  Hey
</a>
`;

exports[`styled works with \`styled.component\` 1`] = `
<a
  style={
    Object {
      "color": "blue",
      "fontWeight": "bold",
    }
  }
>
  Hey
</a>
`;
